shoes-swt/lib/shoes/swt/common/painter.rb

```diff
@@ -30,22 +30,17 @@ def paint_control(event)
         end
 
         def paint_object(graphics_context)
-          cap = LINECAP[@obj.dsl.style[:cap]]
+          dsl = @obj.dsl
+          cap = LINECAP[dsl.style[:cap]]
           graphics_context.set_line_cap(cap) if cap
           graphics_context.set_transform(@obj.transform)
-          obj = @obj.dsl
-          case obj
-            when ::Shoes::Oval, ::Shoes::Rect
-              set_rotate graphics_context, obj.rotate,
-                         obj.element_left + obj.element_width / 2.0,
-                         obj.element_top + obj.element_height / 2.0 do
-                fill graphics_context if fill_setup(graphics_context)
-                draw graphics_context if draw_setup(graphics_context)
-              end
-            else
-              fill graphics_context if fill_setup(graphics_context)
-              draw graphics_context if draw_setup(graphics_context)
-          end
+
+          rotation = rotate(graphics_context, dsl) if dsl.needs_rotate?
+
+          fill graphics_context if fill_setup(graphics_context)
+          draw graphics_context if draw_setup(graphics_context)
+        ensure
+          undo_rotate(graphics_context, dsl, rotation) if rotation
```

Comment:
not sure why we need this `undo_rotate` ? Why is it needed now and not before? I'm sure there is a reason, I just don't get it :)

But yeah in this case it's probably better to get with the solution you had before (maybe + pulling out the `fill`/`draw`)